crates/subspace-networking/src/utils.rs

```diff
@@ -95,3 +97,150 @@ pub(crate) fn convert_multiaddresses(addresses: Vec<Multiaddr>) -> Vec<PeerAddre
+struct SemShared {
+    // The tuple holds (current usage, current max capacity)
+    current: Mutex<SemState>,
+
+    // To signal waiters for permits to be available
+    notify: Notify,
+}
+
+// Current state.
+#[derive(Debug)]
+struct SemState {
+    // The current capacity
+    capacity: usize,
+
+    // The current outstanding permits
+    usage: usize,
+}
+
+impl SemShared {
+    fn new(capacity: usize) -> Self {
+        Self {
+            current: Mutex::new(SemState { capacity, usage: 0 }),
+            notify: Notify::new(),
+        }
+    }
+
+    // Allocates a permit if available.
+    // Returns true if a permit was allocated, false otherwise
+    async fn alloc_one(&self) -> bool {
+        let mut current = self.current.lock().await;
+        if current.usage < current.capacity {
+            current.usage += 1;
+            true
+        } else {
+            false
+        }
+    }
+
+    // Returns a permit back to the free pool, notifies waiters if needed.
+    async fn free_one(&self) {
+        let should_notify = {
+            let mut current = self.current.lock().await;
+            if current.usage > 0 {
+                current.usage -= 1;
+            } else {
+                panic!(
+                    "SemShared::free_one(): invalid free, current = {:?}",
+                    current
+                );
+            }
+
+            // Notify only if usage fell below the current capacity.
+            // For example: if the previous capacity was 100, and current capacity
+            // is 50, this will wait for usage to fall below 50 before any waiters
+            // are notified.
+            current.usage < current.capacity
+        };
+        if should_notify {
+            self.notify.notify_waiters();
+        }
+    }
+
+    // Expands the max capacity by delta, and notifies any waiters of the newly available
+    // free permits.
+    async fn expand(&self, delta: usize) {
+        {
+            let mut current = self.current.lock().await;
+            current.capacity += delta;
+        }
+        self.notify.notify_waiters();
+    }
+
+    // Shrinks the max capacity by delta
+    async fn shrink(&self, delta: usize) -> Result<(), String> {
+        let mut current = self.current.lock().await;
+        if current.capacity > delta {
+            current.capacity -= delta;
+            Ok(())
+        } else {
+            Err(format!(
+                "SemShared::::shrink(): invalid delta = {}, current = {:?}",
+                delta, current
+            ))
+        }
+    }
+}
+
+impl ResizableSemaphore {
+    pub(crate) fn new(capacity: usize) -> Self {
+        Self(Arc::new(SemShared::new(capacity)))
+    }
+
+    // Acquires a permit. Waits until a permit is available.
+    pub(crate) async fn acquire(&self) -> ResizableSemaphorePermit {
+        loop {
+            if self.0.alloc_one().await {
+                break;
+            }
+            self.0.notify.notified().await;
```

Comment:
This is racy. There is a time gap between first `.await` where we inspect `Mutex` contents and second, where in case `SemState` doesn't have free permits waiting is necessary. However, `SemState` can and will change between those two `.await` points under rare conditions and if nothing else frees permits in the meantime, `self.0.notify.notified().await` will be stuck indefinitely. This will be very hard to debug when it does happen.

Reply:
yes, this is the intended behavior. This is the same behavior with how semaphores(and condition variables) work as in general. Example scenario: all permits are taken, several threads are waiting for a permit. The order in which the waiters will receive the permit (as things get freed) is non-deterministic. I don't think they enforce any FIFO behavior for the waiters

But you are right, we can possibly add a warning message if the num waits exceeds a threshold. This might even help us catch a real bug!

Reply:
Thinking more about this, supporting a FIFO (or more deterministic behavior for waiters) is feasible on top of these changes, with some extra complexity. Pls let me know what you think

Reply:
The issue is that wait can be even forever if no permits are dropped. We have an awkward situation where in pathological case we do have slots for more permits, but we're not issuing any ever.

Reply:
I see, let me chat with you about this and post a summary here (to avoid too many messages). Basically, I want to understand this scenario, and why we won't see this with a standard semaphore.